src/raftstore/store/store.rs

```
@@ -483,26 +488,48 @@ impl<T: Transport, C: PdClient> Store<T, C> {
-                    Err(e) => {
-                        // TODO: should we panic or shutdown the store?
-                        error!("handle raft ready at region {} err: {:?}", region_id, e);
-                        return Err(e);
+        let mut send_cnt = 0;
+        let (tx, rx) = mpsc::channel();
+        for &region_id in &ids {
+            if let Some(mut peer) = self.region_peers.remove(&region_id) {
+                let sender = tx.clone();
+                let trans = self.trans.clone();
+                self.apply_pool.execute(move || {
+                    let res = peer.handle_raft_ready(&trans);
+                    if let Err(SendError((peer, _))) = sender.send((peer, res)) {
+                        // TODO: retry here.
+                        error!("failed to report ready result of [{}] {}, will retry",
+                               peer.region_id(),
+                               peer.peer_id());
                     }
-                    Ok(ready) => ready_result = ready,
-                }
+                });
+                send_cnt += 1;
             }
+        }
 
-            if let Some(ready_result) = ready_result {
-                if let Err(e) = self.on_ready_result(region_id, ready_result) {
-                    error!("handle raft ready result at region {} err: {:?}",
-                           region_id,
-                           e);
-                    return Err(e);
+        let mut has_err = false;
+        for _ in 0..send_cnt {
+            // if err, means we lost some peer.
+            let (peer, res) = rx.recv().expect("recv ready result can't failed");
+
+            let region_id = peer.region_id();
+            self.region_peers.insert(region_id, peer);
+
+            match res {
+                Err(e) => {
+                    // TODO: should we panic or shutdown the store?
+                    error!("handle raft ready at region {} err: {:?}", region_id, e);
+                    has_err = true;
```

I think it's better to propagate the error.